### PR TITLE
Wrap `create_nonmatching_meshes_interpolation_data` in Python

### LINKED ISSUE #3021: Wrap non-matching interpolation in Python
### Summarize the issue

Current Python interface is rather rough:
https://github.com/FEniCS/dolfinx/blob/f0054ad8061cac4285e37354b6b48d0a0b3cd76d/python/test/unit/fem/test_interpolation.py#L737-L740

### How to reproduce the bug

Found in the `test_interpolation.py`

### Minimal Example (Python)

```Python
u1.interpolate(u0, nmm_interpolation_data=create_nonmatching_meshes_interpolation_data(
        u1.function_space.mesh._cpp_object,
        u1.function_space.element,
        u0.function_space.mesh._cpp_object, padding=padding))
```

### Output (Python)

_No response_

### Version

main branch

### DOLFINx git commit

_No response_

### Installation

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by mscroggs**
Resolves #3021

### CODE REVIEW

File: python/dolfinx/fem/__init__.py [hunk 1]
```diff
@@ -34,6 +46,30 @@ def create_sparsity_pattern(a: Form):
     return _create_sparsity_pattern(a._cpp_object)
 
 
+def create_nonmatching_meshes_interpolation_data(
+    mesh_to: typing.Union[_Mesh, typing.Union[_Geometry_float64, _Geometry_float32]],
```
Comment (garth-wells) [inline python/dolfinx/fem/__init__.py:50]: Is the nested union required?

File: python/dolfinx/fem/__init__.py [hunk 2]
```diff
@@ -34,6 +46,30 @@ def create_sparsity_pattern(a: Form):
...
+    element: typing.Union[_FiniteElement_float32, _FiniteElement_float64],
+    mesh_from: _Mesh, cells: typing.Optional[npt.NDArray[np.int32]] = None, padding: float = 1e-14
+):
+    """Generate data needed to interpolate discrete functions across different meshes.
+
+    Args:
+        mesh_to: Mesh or geometry to interpolate into
```
Comment (garth-wells) [inline python/dolfinx/fem/__init__.py:57]: Interpolation into a function space makes sense, but it's not clear to me what interpolation in a mesh or geometry means.

File: python/dolfinx/fem/__init__.py [hunk 3]
```diff
@@ -34,6 +46,30 @@ def create_sparsity_pattern(a: Form):
...
+        element: Element of the space to interpolate into
+        mesh_from: Mesh of the function to interpolate from
+        cells: Indices of the cells in the destination mesh on which to interpolate.
+        padding: Absolute padding of bounding boxes of all entities on mesh_to
+    """
+    if cells is None:
+        assert isinstance(mesh_to, _Mesh)
```
Comment (garth-wells) [inline python/dolfinx/fem/__init__.py:64]: I don't think the assert is necessary - just let an exception be raised.

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: python/dolfinx/fem/__init__.py (hunk 2 above)
Comment (mscroggs) [reply python/dolfinx/fem/__init__.py]: I've attempted to make this clearer (based on my incomplete understanding of what the C++ function is doing)

File: python/dolfinx/fem/__init__.py [hunk 4]
```diff
@@ -57,6 +66,35 @@ def create_sparsity_pattern(a: Form):
     return _create_sparsity_pattern(a._cpp_object)
 
 
+def create_nonmatching_meshes_interpolation_data(
+    mesh_to: typing.Union[_Mesh, _Geometry_float64, _Geometry_float32],
+    element: typing.Union[_FiniteElement_float32, _FiniteElement_float64],
+    mesh_from: _Mesh,
+    cells: typing.Optional[npt.NDArray[np.int32]] = None,
+    padding: float = 1e-14,
+) -> tuple[list[np.int32], list[np.int32], list[np.floating], list[np.int32]]:
```
Comment (mscroggs) [inline python/dolfinx/fem/__init__.py:75]: It would be nice to document what each of these lists contains. @jorgensd Any ideas?

Comment (garth-wells) [reply python/dolfinx/fem/__init__.py:75]: Perhaps used a `namedtuple`?

**Comment by jorgensd** [comment]
I cannot reproduce the `rocky` CI failure locally on my system using docker. Does anyone use RedHat and have to to try this?

File: python/dolfinx/fem/function.py [hunk 5]
```diff
@@ -25,6 +25,23 @@
     from dolfinx.mesh import Mesh
 
 
+class PointOwnerShipData(typing.NamedTuple):
```
Comment (mscroggs) [inline python/dolfinx/fem/function.py:28]:
```suggestion
class PointOwnershipData(typing.NamedTuple):
```

File: python/dolfinx/fem/__init__.py [hunk 6]
```diff
@@ -38,6 +46,24 @@
     FunctionSpace,
     functionspace,
 )
+from dolfinx.mesh import Mesh as _Mesh
+
+
+class PointOwnerShipData(typing.NamedTuple):
+    """
+    Convenience class for storing data related to the ownership of points.
```
Comment (mscroggs) [inline python/dolfinx/fem/__init__.py:54]: doc test should fail here?

**Pull request merged**